<a href="https://colab.research.google.com/github/justthea/cs598-dl4h-uncertainty-qa-ehr/blob/main/Companion_Notebook_to_Reproducibility_Study_of_Uncertainty_Aware_Text_to_Program.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Companion Notebook to Reproducibility Study of "Uncertainty-Aware Text-to-Program"

**Author:** Jingtong Xing

**Email:** jxing11@illinois.edu

**Course:** CS598 Deep Learning for Healthcare

## 0. Project Setup and Dependencies

This notebook attempts to reproduce core components of the paper "Uncertainty-Aware Text-to-Program" using the MIMIC-IV clinical database (demo version for portability). We will cover:

1.  Environment and MIMIC-IV demo data setup.
2.  Knowledge graph construction (using DuckDB with MIMIC-IV demo CSVs).
3.  Implementation of atomic database operations.
4.  Use of a small, embedded set of synthetic Question-Program pairs for T5 model pre-training.
5.  Pre-training a T5-base model on this synthetic data using PyTorch Lightning.
6.  Evaluating the model via program generation and execution accuracy.
7.  Demonstrating ambiguity detection using Program Inconsistency Score (PIS) and AUROC.
8.  Showing an example of a clarification prompt.

**Important Note for Reviewers:** This notebook is designed to be self-contained and runnable in Google Colab. It uses the MIMIC-IV demo dataset, which is publicly available and downloaded directly in the notebook. The synthetic Question-Program pairs used for training are embedded within the notebook itself for ease of execution.

In [1]:
!pip install -q duckdb pandas torch torchvision torchaudio torchtext transformers sentencepiece pytorch-lightning tqdm scikit-learn Levenshtein

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 121.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 98.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 61.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 107.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 91.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.1/823.1 kB 62.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

## 1. Environment Initialization and Data Paths

Import necessary libraries and define key paths for our project within the Colab environment.

In [2]:
import os
import duckdb
import pandas as pd
import time
import json
import math
import random
import numpy as np
from pathlib import Path
import re
import glob
from tqdm.auto import tqdm
import itertools

import torch
from torch.utils.data import Dataset, DataLoader, random_split
from transformers import T5Tokenizer, T5ForConditionalGeneration, get_linear_schedule_with_warmup
from torch.optim import AdamW
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelCheckpoint, EarlyStopping
from pytorch_lightning.loggers import CSVLogger
from sklearn.metrics import roc_auc_score, roc_curve
import Levenshtein

# Project paths in Colab
COLAB_PROJECT_ROOT = Path('/content/text2program_mimic_repro')
MIMIC_DEMO_DOWNLOAD_DIR = COLAB_PROJECT_ROOT / 'mimic_iv_demo_data'
MIMIC_DEMO_CSV_ROOT = MIMIC_DEMO_DOWNLOAD_DIR / 'physionet.org/files/mimic-iv-demo/2.2/'
OUTPUTS_DIR = COLAB_PROJECT_ROOT / 'outputs'
CHECKPOINTS_DIR = OUTPUTS_DIR / 'checkpoints'
LOGS_DIR = COLAB_PROJECT_ROOT / 'lightning_logs'

os.makedirs(COLAB_PROJECT_ROOT, exist_ok=True)
os.makedirs(MIMIC_DEMO_DOWNLOAD_DIR, exist_ok=True)
os.makedirs(OUTPUTS_DIR, exist_ok=True)
os.makedirs(CHECKPOINTS_DIR, exist_ok=True)
os.makedirs(LOGS_DIR, exist_ok=True)

print(f"Project Root: {COLAB_PROJECT_ROOT}")
print(f"MIMIC Demo Download Dir: {MIMIC_DEMO_DOWNLOAD_DIR}")
print(f"Expected MIMIC Demo CSV Root (after download): {MIMIC_DEMO_CSV_ROOT}")
print(f"Outputs Dir: {OUTPUTS_DIR}")

Project Root: /content/text2program_mimic_repro
MIMIC Demo Download Dir: /content/text2program_mimic_repro/mimic_iv_demo_data
Expected MIMIC Demo CSV Root (after download): /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2
Outputs Dir: /content/text2program_mimic_repro/outputs


## 2. Download MIMIC-IV Demo Dataset

We'll use `wget` to download the MIMIC-IV v2.2 demo dataset directly into our Colab environment. This dataset is publicly available from PhysioNet.

In [3]:
print(f"Downloading MIMIC-IV v2.2 demo dataset to {MIMIC_DEMO_DOWNLOAD_DIR}...")
!wget -r -N -c -np -P {MIMIC_DEMO_DOWNLOAD_DIR} https://physionet.org/files/mimic-iv-demo/2.2/

print("\nVerifying download...")
expected_patients_file = MIMIC_DEMO_CSV_ROOT / 'hosp/patients.csv.gz'
if expected_patients_file.exists():
    print(f"Successfully found patients file: {expected_patients_file}")
    print("Listing contents of the demo CSV root directory:")
    for item in os.listdir(MIMIC_DEMO_CSV_ROOT):
        print(f"  - {item}")
    print("Listing contents of the 'hosp' subdirectory:")
    for item in os.listdir(MIMIC_DEMO_CSV_ROOT / 'hosp'):
        print(f"  - {item}")
else:
    print(f"ERROR: Patients file not found at {expected_patients_file}. Please check the download.")

--2025-05-07 20:50:45--  https://physionet.org/files/mimic-iv-demo/2.2/
Resolving physionet.org (physionet.org)... 18.18.42.54
Connecting to physionet.org (physionet.org)|18.18.42.54|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘/content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/index.html’

physionet.org/files     [ <=>                ]     916  --.-KB/s    in 0s      

Last-modified header missing -- time-stamps turned off.
2025-05-07 20:50:46 (176 MB/s) - ‘/content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/index.html’ saved [916]

Loading robots.txt; please ignore errors.
--2025-05-07 20:50:46--  https://physionet.org/robots.txt
Reusing existing connection to physionet.org:443.
HTTP request sent, awaiting response... 200 OK
Length: 22 [text/plain]
Saving to: ‘/content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/robots.txt’

phy

## 3. Knowledge Graph Setup (DuckDB)

We'll use DuckDB, an in-memory analytical database, to act as our knowledge graph. We'll register the key MIMIC-IV demo CSV files as views in DuckDB.

In [4]:
db_connection = None
try:
    db_connection = duckdb.connect(database=':memory:', read_only=False)
    print("DuckDB connection established.")

    # define tables to register based on MIMIC-IV v2.2 demo structure
    tables_for_registration = {
        'patients': 'hosp/patients.csv.gz',
        'admissions': 'hosp/admissions.csv.gz',
        'diagnoses_icd': 'hosp/diagnoses_icd.csv.gz',
        'd_icd_diagnoses': 'hosp/d_icd_diagnoses.csv.gz',
        'labevents': 'hosp/labevents.csv.gz',
        'd_labitems': 'hosp/d_labitems.csv.gz'
        # note: Add other tables if needed by atomic operations, these are the
        # minimal required for running this notebook with demo data
    }

    registered_db_views = []
    print(f"Attempting to register tables from: {MIMIC_DEMO_CSV_ROOT}")
    for view_name, rel_file_path in tables_for_registration.items():
        full_file_path = MIMIC_DEMO_CSV_ROOT / rel_file_path
        if full_file_path.exists():
            try:
                db_connection.execute(f"CREATE VIEW {view_name} AS SELECT * FROM read_csv_auto('{str(full_file_path)}')")
                registered_db_views.append(view_name)
                print(f"  Successfully registered view: {view_name} from {full_file_path}")
            except Exception as e:
                print(f"  ERROR registering view {view_name} from {full_file_path}: {e}")
        else:
            print(f"  WARNING: File not found for view {view_name}: {full_file_path}")

    print("\nRegistered views in DuckDB:")
    duckdb_tables = db_connection.execute("SHOW TABLES").fetchall()
    for table_tuple in duckdb_tables:
        print(f"  - {table_tuple[0]}")

    # quick test query for sanity check
    if 'patients' in registered_db_views:
        print("\nTesting query on 'patients' view:")
        sample_patients = db_connection.execute("SELECT subject_id, gender, anchor_age FROM patients LIMIT 3").fetchdf()
        print(sample_patients)
    else:
        print("\nSkipping test query as 'patients' view was not registered.")

except Exception as e:
    print(f"An error occurred during DuckDB setup: {e}")
print("DuckDB setup cell finished.")

DuckDB connection established.
Attempting to register tables from: /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2
  Successfully registered view: patients from /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/hosp/patients.csv.gz
  Successfully registered view: admissions from /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/hosp/admissions.csv.gz
  Successfully registered view: diagnoses_icd from /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/hosp/diagnoses_icd.csv.gz
  Successfully registered view: d_icd_diagnoses from /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/hosp/d_icd_diagnoses.csv.gz
  Successfully registered view: labevents from /content/text2program_mimic_repro/mimic_iv_demo_data/physionet.org/files/mimic-iv-demo/2.2/hosp/labevents.csv.gz
  Successfully registe

## 4. Embedded Synthetic Question-Program Pairs

For this Colab notebook, instead of generating synthetic pairs from scratch (which requires the full MIMIC-IV dataset for meaningful diversity and complex templates), we will embed a pre-generated set of 56 diverse pairs. These pairs were created using a more extensive templating system with the full MIMIC-IV dataset and are representative of the types of questions and programs the model should learn. This allows the notebook to be self-contained and focus on the model training and evaluation aspects with the demo data where applicable for execution.

In [5]:
embedded_synthetic_pairs = [
    {'question': "How many male patients had an anchor age equal to 70?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Other and unspecified malignant neoplasms of lymphoid, hematopoietic and related tissue' (code: C96, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='C96', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many admissions were there for female patients?", 'program': "count_entset(gen_entset_down(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', target_table='admissions', target_table_join_col='subject_id', desired_target_entity_col='hadm_id'), entity_col_name='hadm_id')"},
    {'question': "How many lab entries exist for Homocysteine (itemid 50945)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='50945', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "Count all female patients.", 'program': "count_entset(gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients had an anchor age less than or equal to 30?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<=', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Unspecified acute lower respiratory infection' (code: J22, version: 9)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='J22', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for Eosinophil Count (itemid 51133)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='51133', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "Count all patients.", 'program': "count_entset(gen_entset_all(table_name='patients', entity_column='subject_id'), entity_col_name='subject_id')"},
    {'question': "What is the maximum anchor age of any patient?", 'program': "max_litset(gen_litset(table_name='patients', literal_column='anchor_age'), literal_col_name='anchor_age')"},
    {'question': "How many female patients had an anchor age greater than 50?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='>', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many admissions were there for male patients?", 'program': "count_entset(gen_entset_down(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', target_table='admissions', target_table_join_col='subject_id', desired_target_entity_col='hadm_id'), entity_col_name='hadm_id')"},
    {'question': "How many lab entries exist for Bilirubin, Total (itemid 53089)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='53089', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "Count all male patients.", 'program': "count_entset(gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients had an anchor age less than 30?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Type 2 diabetes mellitus with foot ulcer' (code: E11621, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='E11621', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many female patients had an anchor age equal to 70?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "What is the minimum anchor age of any patient?", 'program': "min_litset(gen_litset(table_name='patients', literal_column='anchor_age'), literal_col_name='anchor_age')"},
    {'question': "How many lab entries exist for Monocytes (itemid 52075)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='52075', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many female patients had an anchor age less than or equal to 50?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<=', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients were older than 30 years based on anchor age?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='>', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Poisoning by antiparkinsonism drugs and other central muscle-tone depressants, assault, sequela' (code: T428X3S, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='T428X3S', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for Hepatitis B Virus E Antigen (itemid 51650)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='51650', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many female patients had an anchor age less than 70?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients had an anchor age equal to 30?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Fracture of corpus callosum' (code: S061X7, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='S061X7', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for C-Reactive Protein (itemid 50885)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='50885', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many male patients were older than 70 years based on anchor age?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='>', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for LDL Cholesterol (itemid 51638)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='51638', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many female patients had an anchor age less than or equal to 70?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<=', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients had an anchor age less than or equal to 70?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<=', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Nondisplaced fracture of greater trochanter of right femur, subsequent encounter for open fracture type I or II with routine healing' (code: S72121E, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='S72121E', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many female patients had an anchor age less than 30?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients had an anchor age less than 70?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for Prolactin (itemid 51704)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='51704', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many female patients had an anchor age equal to 30?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Stress incontinence (female) (male)' (code: N393, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='N393', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Acute appendicitis with perforation, localized peritonitis, and gangrene, with abscess' (code: K3533, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='K3533', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Post-traumatic epilepsy, not intractable, with status epilepticus' (code: G40501, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='G40501', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many male patients are there?", 'program': "count_entset(gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Displaced fracture of medial condyle of left tibia, subsequent encounter for closed fracture with malunion' (code: S82125K, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='S82125K', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many female patients were older than 30 years based on anchor age?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='>', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Displaced fracture of base of neck of right femur, subsequent encounter for closed fracture with nonunion' (code: S72021M, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='S72021M', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Parkinson's disease' (code: G20, version: 9)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='G20', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for Albumin (itemid 50862)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='50862', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many patients had a diagnosis of 'Superficial foreign body, right thigh, initial encounter' (code: S70251A, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='S70251A', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many female patients were older than 70 years based on anchor age?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='>', comparison_value=70, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for Renin (itemid 51715)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='51715', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many lab entries exist for HIV Viral Load Ct (itemid 53163)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='53163', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many male patients had an anchor age equal to 50?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many male patients were older than 50 years based on anchor age?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='>', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many lab entries exist for Folate (itemid 50927)?", 'program': "count_entset(gen_entset_equal(table_name='labevents', column_name='itemid', value='50927', entity_column='row_id'), entity_col_name='row_id')"},
    {'question': "How many female patients had an anchor age equal to 50?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many patients had a diagnosis of 'Other specified injury of unspecified blood vessel of thorax, sequela' (code: S2589XS, version: 10)?", 'program': "count_entset(gen_entset_equal(table_name='diagnoses_icd', column_name='icd_code', value='S2589XS', entity_column='subject_id'),entity_col_name='subject_id')"},
    {'question': "How many male patients had an anchor age less than 50?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"},
    {'question': "How many female patients had an anchor age less than 50?", 'program': "count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<', comparison_value=50, attribute_table_join_col='subject_id'), entity_col_name='subject_id')"}
]

print(f"Loaded {len(embedded_synthetic_pairs)} embedded synthetic question-program pairs.")
print("Sample embedded pair:")
for i, pair_item in enumerate(random.sample(embedded_synthetic_pairs, min(3, len(embedded_synthetic_pairs)))):
    print(f"  --- Sample {i+1} ---")
    print(f"    Q: {pair_item['question']}")
    print(f"    P: {pair_item['program']}")

Loaded 56 embedded synthetic question-program pairs.
Sample embedded pair:
  --- Sample 1 ---
    Q: How many male patients had an anchor age less than or equal to 30?
    P: count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='<=', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')
  --- Sample 2 ---
    Q: How many female patients had an anchor age equal to 30?
    P: count_entset(filter_entset_comparison(source_entset_df=gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), source_entity_col='subject_id', attribute_table='patients', attribute_col='anchor_age', comparison_operator='=', comparison_value=30, attribute_table_join_col='subject_id'), entity_col_name='subject_id')
  --- Sample 

## 5. Define Atomic Operations

These are the core functions that interact with the DuckDB database to retrieve and manipulate data, forming the building blocks of our programs.

In [6]:
def execute_duckdb_query(query_str):
    if 'db_connection' not in globals() or db_connection is None:
        raise ConnectionError("DuckDB connection not available. Ensure Cell [3] has run successfully.")
    try:
        return db_connection.execute(query_str).fetchdf()
    except Exception as e:
        # print(f"Error executing query: {query_str}\nError: {e}")
        raise e

def gen_entset_all(table_name, entity_column):
    query = f"SELECT DISTINCT {entity_column} FROM {table_name}"
    return execute_duckdb_query(query)

def gen_entset_equal(table_name, column_name, value, entity_column):
    if isinstance(value, str):
        db_val = f"'{value}'"
    else:
        db_val = value
    query = f"SELECT DISTINCT {entity_column} FROM {table_name} WHERE {column_name} = {db_val}"
    return execute_duckdb_query(query)

def count_entset(entity_set_df, entity_col_name):
    if not isinstance(entity_set_df, pd.DataFrame):
        # print(f"Warning: count_entset received non-DataFrame: {type(entity_set_df)}")
        return 0
    return entity_set_df[entity_col_name].nunique()

def gen_entset_down(source_entset_df, source_entity_col, target_table, target_table_join_col, desired_target_entity_col):
    if not isinstance(source_entset_df, pd.DataFrame) or source_entset_df.empty:
        return pd.DataFrame(columns=[desired_target_entity_col])

    source_ids = tuple(source_entset_df[source_entity_col].unique())
    if not source_ids:
        return pd.DataFrame(columns=[desired_target_entity_col])

    # Create a temporary view for the source entity set to use in join
    db_connection.register('temp_source_entset_for_down', source_entset_df)

    query = f"""
    SELECT DISTINCT t.{desired_target_entity_col}
    FROM {target_table} t
    JOIN temp_source_entset_for_down s ON t.{target_table_join_col} = s.{source_entity_col}
    """
    result_df = execute_duckdb_query(query)
    db_connection.unregister('temp_source_entset_for_down') # Clean up temp view
    return result_df

def gen_litset(table_name, literal_column, entity_set_df=None, entity_column_for_filter=None):
    if entity_set_df is not None and not entity_set_df.empty and entity_column_for_filter is not None:
        filter_ids = tuple(entity_set_df[entity_column_for_filter].unique())
        if not filter_ids:
             return pd.DataFrame(columns=[literal_column])
        db_connection.register('temp_filter_entset_for_litset', entity_set_df)
        query = f"""
        SELECT DISTINCT T1.{literal_column}
        FROM {table_name} T1
        JOIN temp_filter_entset_for_litset T2 ON T1.{entity_column_for_filter} = T2.{entity_column_for_filter}
        """
        result_df = execute_duckdb_query(query)
        db_connection.unregister('temp_filter_entset_for_litset')
        return result_df
    else:
        query = f"SELECT DISTINCT {literal_column} FROM {table_name}"
        return execute_duckdb_query(query)

def max_litset(literal_set_df, literal_col_name):
    if not isinstance(literal_set_df, pd.DataFrame) or literal_set_df.empty:
        return None
    return literal_set_df[literal_col_name].max()

def min_litset(literal_set_df, literal_col_name):
    if not isinstance(literal_set_df, pd.DataFrame) or literal_set_df.empty:
        return None
    return literal_set_df[literal_col_name].min()

def intersect_entsets(entset1_df, entset2_df, common_entity_col='subject_id'): # Assuming subject_id is common
    if not isinstance(entset1_df, pd.DataFrame) or not isinstance(entset2_df, pd.DataFrame):
        # print("Warning: Intersect received non-DataFrame.")
        return pd.DataFrame(columns=[common_entity_col])
    if entset1_df.empty or entset2_df.empty:
        return pd.DataFrame(columns=[common_entity_col])

    # Ensure the common entity column exists in both DataFrames
    if common_entity_col not in entset1_df.columns or common_entity_col not in entset2_df.columns:
        # print(f"Warning: Common entity column '{common_entity_col}' not in one or both DataFrames for intersection.")
        return pd.DataFrame(columns=[common_entity_col])

    # Perform intersection using pandas merge, then select and drop duplicates for the common_entity_col
    merged_df = pd.merge(entset1_df[[common_entity_col]].drop_duplicates(),
                         entset2_df[[common_entity_col]].drop_duplicates(),
                         on=common_entity_col,
                         how='inner')
    return merged_df

def filter_entset_comparison(source_entset_df, source_entity_col, attribute_table, attribute_col, comparison_operator, comparison_value, attribute_table_join_col):
    if not isinstance(source_entset_df, pd.DataFrame) or source_entset_df.empty:
        return pd.DataFrame(columns=[source_entity_col])

    valid_ops = ['>', '<', '>=', '<=', '=', '!=']
    if comparison_operator not in valid_ops:
        raise ValueError(f"Invalid comparison operator: {comparison_operator}. Must be one of {valid_ops}")

    db_connection.register('temp_source_entset_for_filter', source_entset_df)

    # Ensure comparison_value is properly quoted if it's a string for the SQL query
    sql_comp_value = f"'{comparison_value}'" if isinstance(comparison_value, str) else comparison_value

    query = f"""
    SELECT DISTINCT s.{source_entity_col}
    FROM temp_source_entset_for_filter s
    JOIN {attribute_table} attr ON s.{source_entity_col} = attr.{attribute_table_join_col}
    WHERE attr.{attribute_col} {comparison_operator} {sql_comp_value}
    """

    result_df = execute_duckdb_query(query)
    db_connection.unregister('temp_source_entset_for_filter') # Clean up temp view
    return result_df

print("Atomic operations defined.")

Atomic operations defined.


## 6. Test Atomic Operations (Basic Sanity Checks)

Perform a few simple tests to ensure the atomic operations are working as expected with the demo data.

In [7]:
print("Testing atomic operations...")
try:
    # Test 1: gen_entset_equal & count_entset
    female_patients_df = gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id')
    num_female_patients = count_entset(female_patients_df, 'subject_id')
    print(f"Number of unique female demo patients: {num_female_patients}")
    # print(female_patients_df.head())

    # Test 2: gen_entset_down
    if not female_patients_df.empty:
        admissions_female_df = gen_entset_down(female_patients_df, 'subject_id', 'admissions', 'subject_id', 'hadm_id')
        num_admissions_female = count_entset(admissions_female_df, 'hadm_id')
        print(f"Number of unique admissions for female demo patients: {num_admissions_female}")
        # print(admissions_female_df.head())
    else:
        print("Skipping gen_entset_down test as no female patients found or error occurred.")

    # Test 3: gen_litset and min/max_litset
    anchor_years_df = gen_litset(table_name='patients', literal_column='anchor_year')
    if not anchor_years_df.empty:
        min_anchor_year = min_litset(anchor_years_df, 'anchor_year')
        max_anchor_year = max_litset(anchor_years_df, 'anchor_year')
        print(f"Min anchor year in demo: {min_anchor_year}, Max anchor year in demo: {max_anchor_year}")
    else:
        print("Skipping min/max_litset test as no anchor years found or error occurred.")

    # Test 4: filter_entset_comparison
    all_patients_df = gen_entset_all(table_name='patients', entity_column='subject_id')
    patients_anchor_age_gt_60_df = filter_entset_comparison(
        source_entset_df=all_patients_df,
        source_entity_col='subject_id',
        attribute_table='patients',
        attribute_col='anchor_age',
        comparison_operator='>',
        comparison_value=60,
        attribute_table_join_col='subject_id'
    )
    num_patients_anchor_age_gt_60 = count_entset(patients_anchor_age_gt_60_df, 'subject_id')
    print(f"Number of demo patients with anchor_age > 60: {num_patients_anchor_age_gt_60}")

    print("Atomic operation tests completed.")
except Exception as e:
    print(f"An error occurred during atomic operation testing: {e}")

Testing atomic operations...
Number of unique female demo patients: 43
Number of unique admissions for female demo patients: 133
Min anchor year in demo: 2110, Max anchor year in demo: 2201
Number of demo patients with anchor_age > 60: 55
Atomic operation tests completed.


## 7. T5 Model Pre-training Setup

Define hyperparameters, tokenizer, and the PyTorch Dataset for loading our synthetic Question-Program pairs.

In [8]:
MODEL_NAME = 't5-base'
LEARNING_RATE = 1e-4 # From Table 4 of paper
ADAM_EPSILON = 1e-8  # Default for AdamW
WEIGHT_DECAY = 0.0   # Default for AdamW
TRAIN_BATCH_SIZE = 2 # Reduced for demo; paper uses 32 or 64
VALID_BATCH_SIZE = 2 # Reduced for demo
MAX_EPOCHS = 3       # Reduced for demo; paper uses 10-20 for pre-training
INPUT_MAX_LEN = 256  # Max length for tokenized questions (paper: 256)
OUTPUT_MAX_LEN = 128 # Max length for tokenized programs (paper: 128)
SEED = 42
pl.seed_everything(SEED)

ACCELERATOR_TYPE = 'gpu' if torch.cuda.is_available() else ('mps' if torch.backends.mps.is_available() else 'cpu')
NUM_DEVICES = 1

print(f"Using accelerator: {ACCELERATOR_TYPE}, devices: {NUM_DEVICES}")
EFFECTIVE_DEVICE = ACCELERATOR_TYPE if ACCELERATOR_TYPE != 'cpu' else 'cpu' # PyTorch Lightning handles 'gpu' vs 'mps'
if ACCELERATOR_TYPE == 'gpu':
    EFFECTIVE_DEVICE = 'cuda'
elif ACCELERATOR_TYPE == 'mps':
     EFFECTIVE_DEVICE = 'mps'
print(f"PyTorch device will be effectively: {EFFECTIVE_DEVICE}")

try:
    text_tokenizer = T5Tokenizer.from_pretrained(MODEL_NAME, legacy=False)
    print(f"Tokenizer for '{MODEL_NAME}' loaded successfully.")
except Exception as e:
    print(f"Error loading tokenizer: {e}")
    text_tokenizer = None # Ensure it's defined even on error

class NLQProgramDataset(Dataset):
    def __init__(self, data_pairs, tokenizer, input_max_len, output_max_len):
        self.data_pairs = data_pairs
        self.tokenizer = tokenizer
        self.input_max_len = input_max_len
        self.output_max_len = output_max_len

    def __len__(self):
        return len(self.data_pairs)

    def __getitem__(self, idx):
        pair = self.data_pairs[idx]
        question = "translate English to Program: " + pair['question'] # Task prefix
        program_code = pair['program']

        # tokenize inputs (question)
        input_encoding = self.tokenizer(
            question,
            max_length=self.input_max_len,
            padding='max_length',
            truncation=True,
            return_attention_mask=True,
            add_special_tokens=True,
            return_tensors='pt'
        )

        # tokenize outputs (program code) for labels
        # For T5, the decoder input_ids are created by shifting the labels right
        # <pad> tokens in labels are replaced with -100 to be ignored by the loss function
        labels_encoding = self.tokenizer(
            program_code,
            max_length=self.output_max_len,
            padding='max_length',
            truncation=True,
            return_attention_mask=False, # Not needed for labels in this setup
            add_special_tokens=True,
            return_tensors='pt'
        )

        labels = labels_encoding['input_ids'].squeeze()
        labels[labels == self.tokenizer.pad_token_id] = -100 # Mask padding tokens for loss calculation

        return {
            'input_ids': input_encoding['input_ids'].squeeze(),
            'attention_mask': input_encoding['attention_mask'].squeeze(),
            'labels': labels
        }

# Test the Dataset
if text_tokenizer:
    print("\n--- Testing NLQProgramDataset ---")
    # Use the embedded_synthetic_pairs defined in Cell [4]
    synthetic_dataset_for_test = NLQProgramDataset(embedded_synthetic_pairs, text_tokenizer, INPUT_MAX_LEN, OUTPUT_MAX_LEN)
    print(f"Dataset size: {len(synthetic_dataset_for_test)}")
    if len(synthetic_dataset_for_test) > 0:
        sample_item = synthetic_dataset_for_test[0]
        print("Sample item from dataset:")
        for key, val_tensor in sample_item.items():
            print(f"  {key}: shape {val_tensor.shape}, dtype {val_tensor.dtype}")

        print("\nDecoded sample:")
        decoded_input = text_tokenizer.decode(sample_item['input_ids'], skip_special_tokens=False) # Show special tokens
        # For labels, replace -100 with pad_token_id before decoding for viewability
        temp_labels = sample_item['labels'].clone()
        temp_labels[temp_labels == -100] = text_tokenizer.pad_token_id
        decoded_output = text_tokenizer.decode(temp_labels, skip_special_tokens=False) # Show special tokens
        print(f"  Input (Question): {decoded_input}")
        print(f"  Output (Program): {decoded_output}")
    else:
        print("Dataset is empty, cannot show sample.")
else:
    print("Tokenizer not loaded, skipping NLQProgramDataset test.")

INFO:lightning_fabric.utilities.seed:Seed set to 42


Using accelerator: gpu, devices: 1
PyTorch device will be effectively: cuda


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

Tokenizer for 't5-base' loaded successfully.

--- Testing NLQProgramDataset ---
Dataset size: 56
Sample item from dataset:
  input_ids: shape torch.Size([256]), dtype torch.int64
  attention_mask: shape torch.Size([256]), dtype torch.int64
  labels: shape torch.Size([128]), dtype torch.int64

Decoded sample:
  Input (Question): translate English to Program: How many male patients had an anchor age equal to 70?</s><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pad><pa

## 8. Define PyTorch LightningModule for T5 Fine-tuning

This module encapsulates the T5 model, training loop, validation loop, and optimizer configuration.

In [9]:
class T5FineTuner(pl.LightningModule):
    def __init__(self, model_name_or_path, learning_rate, adam_epsilon, weight_decay,
                 tokenizer_instance, train_data_len, train_batch_sz, max_training_epochs):
        super().__init__()
        self.save_hyperparameters(ignore=['tokenizer_instance']) # tokenizer cannot be pickled
        self.core_model = T5ForConditionalGeneration.from_pretrained(model_name_or_path)
        self.ext_tokenizer = tokenizer_instance # Store tokenizer for generation, not saved in checkpoint by PL directly
        # Store these for scheduler setup
        self.hparams.train_data_len = train_data_len
        self.hparams.train_batch_sz = train_batch_sz
        self.hparams.max_training_epochs = max_training_epochs

    def forward(self, input_ids, attention_mask, decoder_input_ids=None, labels=None):
        return self.core_model(
            input_ids,
            attention_mask=attention_mask,
            decoder_input_ids=decoder_input_ids,
            labels=labels
        )

    def training_step(self, batch_data, batch_idx):
        outputs = self.forward(
            input_ids=batch_data['input_ids'],
            attention_mask=batch_data['attention_mask'],
            labels=batch_data['labels']
        )
        loss = outputs.loss
        self.log('train_loss', loss, prog_bar=True, logger=True, on_step=True, on_epoch=True)
        return loss

    def validation_step(self, batch_data, batch_idx):
        outputs = self.forward(
            input_ids=batch_data['input_ids'],
            attention_mask=batch_data['attention_mask'],
            labels=batch_data['labels']
        )
        loss = outputs.loss
        self.log('val_loss', loss, prog_bar=True, logger=True, on_step=False, on_epoch=True)
        return loss

    def configure_optimizers(self):
        optimizer = AdamW(
            self.parameters(),
            lr=self.hparams.learning_rate,
            eps=self.hparams.adam_epsilon,
            weight_decay=self.hparams.weight_decay
        )

        # Calculate total training steps for the scheduler
        num_gpus_or_devices = self.trainer.num_devices if self.trainer else 1
        effective_batch_size = self.hparams.train_batch_sz * num_gpus_or_devices * self.trainer.accumulate_grad_batches
        total_training_steps = (self.hparams.train_data_len // effective_batch_size) * self.hparams.max_training_epochs

        # print(f"DEBUG: Calculated num_training_steps for scheduler: {total_training_steps}")
        # print(f"  train_data_len: {self.hparams.train_data_len}")
        # print(f"  train_batch_sz: {self.hparams.train_batch_sz}")
        # print(f"  max_training_epochs: {self.hparams.max_training_epochs}")
        # print(f"  num_gpus_or_devices: {num_gpus_or_devices}")
        # print(f"  accumulate_grad_batches: {self.trainer.accumulate_grad_batches}")

        scheduler = get_linear_schedule_with_warmup(
            optimizer,
            num_warmup_steps=0, # since paper does not specify warmup, so 0
            num_training_steps=total_training_steps
        )
        return {"optimizer": optimizer, "lr_scheduler": {"scheduler": scheduler, "interval": "step"}}

# Test Instantiation
if text_tokenizer:
    print("--- Testing T5FineTuner Instantiation ---")
    try:
        # Dummy values for instantiation test
        test_model_module = T5FineTuner(
            MODEL_NAME,
            LEARNING_RATE,
            ADAM_EPSILON,
            WEIGHT_DECAY,
            text_tokenizer,
            train_data_len=len(embedded_synthetic_pairs), # actual length
            train_batch_sz=TRAIN_BATCH_SIZE,
            max_training_epochs=MAX_EPOCHS
        )
        print("T5FineTuner module instantiated successfully.")
    except Exception as e:
        print(f"Error instantiating T5FineTuner: {e}")
else:
    print("Tokenizer not available, skipping T5FineTuner instantiation test.")

--- Testing T5FineTuner Instantiation ---


Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

T5FineTuner module instantiated successfully.


## 9. Create DataLoaders

Prepare training and validation DataLoaders from our embedded synthetic dataset.

In [10]:
train_dataset_main = None
val_dataset_main = None
actual_train_data_size = 0

if text_tokenizer and embedded_synthetic_pairs:
    print("Proceeding to create DataLoaders...")
    full_nlq_dataset = NLQProgramDataset(embedded_synthetic_pairs, text_tokenizer, INPUT_MAX_LEN, OUTPUT_MAX_LEN)
    print(f"Full dataset size: {len(full_nlq_dataset)}")

    # Split data: 80% train, 20% validation (or at least 1 for validation if dataset is too small)
    num_total_samples = len(full_nlq_dataset)
    if num_total_samples < 5: # Arbitrary small number, ensure val_size is at least 1
        num_train_samples = num_total_samples - 1 if num_total_samples > 1 else num_total_samples
        num_val_samples = 1 if num_total_samples > 1 else 0
    else:
        num_train_samples = int(0.8 * num_total_samples)
        num_val_samples = num_total_samples - num_train_samples

    print(f"Attempting to split into {num_train_samples} train and {num_val_samples} validation samples.")

    if num_val_samples == 0 and num_train_samples > 0: # Edge case: all data goes to train, no validation
        train_dataset_main = full_nlq_dataset
        val_dataset_main = None # Or a very small dummy if PL requires it
        print("Warning: Entire dataset used for training, no validation set created due to small size.")
    elif num_val_samples > 0 : # Normal split
        train_dataset_main, val_dataset_main = random_split(full_nlq_dataset, [num_train_samples, num_val_samples])
    else: # Dataset is empty or too small for any split
        train_dataset_main = None
        val_dataset_main = None
        print("Dataset too small for train/val split or empty.")

    if train_dataset_main:
        actual_train_data_size = len(train_dataset_main)
        print(f"Train dataset size: {len(train_dataset_main)}")
        train_loader = DataLoader(train_dataset_main, batch_size=TRAIN_BATCH_SIZE, shuffle=True, num_workers=2) # num_workers > 0 for Colab
        print(f"Train DataLoader created with batch size {TRAIN_BATCH_SIZE}.")

        # Test first batch from train_loader
        print("\n--- Testing train_dl (first batch) ---")
        try:
            first_train_batch = next(iter(train_loader))
            print("Batch loaded successfully.")
            print(f"  Input IDs shape: {first_train_batch['input_ids'].shape}")
            print(f"  Attention Mask shape: {first_train_batch['attention_mask'].shape}")
            print(f"  Labels shape: {first_train_batch['labels'].shape}")
        except Exception as e:
            print(f"Error loading first batch from train_loader: {e}")
    else:
        train_loader = None
        print("Train dataset is empty, Train DataLoader not created.")

    if val_dataset_main:
        print(f"Validation dataset size: {len(val_dataset_main)}")
        val_loader = DataLoader(val_dataset_main, batch_size=VALID_BATCH_SIZE, num_workers=2)
        print(f"Validation DataLoader created with batch size {VALID_BATCH_SIZE}.")
    else:
        val_loader = None
        print("Validation dataset is empty or not created, Validation DataLoader not created.")
else:
    train_loader = None
    val_loader = None
    print("Tokenizer or synthetic pairs not available. DataLoaders not created.")

Proceeding to create DataLoaders...
Full dataset size: 56
Attempting to split into 44 train and 12 validation samples.
Train dataset size: 44
Train DataLoader created with batch size 2.

--- Testing train_dl (first batch) ---
Batch loaded successfully.
  Input IDs shape: torch.Size([2, 256])
  Attention Mask shape: torch.Size([2, 256])
  Labels shape: torch.Size([2, 128])
Validation dataset size: 12
Validation DataLoader created with batch size 2.


## 10. Instantiate Model and Trainer

Set up the `T5FineTuner` model and the PyTorch Lightning `Trainer` with callbacks for checkpointing and early stopping.

In [11]:
model_for_training = None
trainer_instance = None

if text_tokenizer and train_dataset_main:
    print("Proceeding to instantiate Model and Trainer...")
    model_for_training = T5FineTuner(
        MODEL_NAME,
        LEARNING_RATE,
        ADAM_EPSILON,
        WEIGHT_DECAY,
        text_tokenizer,
        train_data_len=actual_train_data_size,
        train_batch_sz=TRAIN_BATCH_SIZE,
        max_training_epochs=MAX_EPOCHS
    )
    print(f"T5FineTuner model instantiated successfully with actual_train_data_size={actual_train_data_size}.")

    # Callbacks/checkpoitns
    checkpoint_callback_obj = ModelCheckpoint(
        dirpath=CHECKPOINTS_DIR,
        filename='t5-text2program-{epoch:02d}-vloss{val_loss:.2f}',
        save_top_k=1,
        verbose=True,
        monitor='val_loss',
        mode='min'
    )
    print(f"ModelCheckpoint callback configured to save to: {CHECKPOINTS_DIR}")

    early_stop_callback_obj = EarlyStopping(
        monitor='val_loss',
        min_delta=0.00,
        patience=3,
        verbose=False,
        mode='min'
    )
    print("EarlyStopping callback configured.")

    csv_train_logger = CSVLogger(save_dir=LOGS_DIR)

    trainer_callbacks_list = [checkpoint_callback_obj]
    if val_loader:
        trainer_callbacks_list.append(early_stop_callback_obj)
        print("EarlyStopping callback will be used.")
    else:
        print("EarlyStopping callback will NOT be used as there is no validation data.")

    trainer_instance = pl.Trainer(
        accelerator=ACCELERATOR_TYPE,
        devices=NUM_DEVICES,
        max_epochs=MAX_EPOCHS,
        callbacks=trainer_callbacks_list,
        logger=csv_train_logger,
        log_every_n_steps=10
    )
    print("PyTorch Lightning Trainer instantiated.")
    print(f"  Accelerator: {ACCELERATOR_TYPE}")
    print(f"  Devices: {trainer_instance.num_devices}")
    print(f"  Max Epochs: {trainer_instance.max_epochs}")
    print(f"  Logger: CSVLogger (logs to {LOGS_DIR})")

else:
    print("Tokenizer or training data not available. Model and Trainer not instantiated.")

Proceeding to instantiate Model and Trainer...


INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs


T5FineTuner model instantiated successfully with actual_train_data_size=44.
ModelCheckpoint callback configured to save to: /content/text2program_mimic_repro/outputs/checkpoints
EarlyStopping callback configured.
EarlyStopping callback will be used.
PyTorch Lightning Trainer instantiated.
  Accelerator: gpu
  Devices: 1
  Max Epochs: 3
  Logger: CSVLogger (logs to /content/text2program_mimic_repro/lightning_logs)


## 11. Start T5 Model Pre-training

Begin the fine-tuning process using the prepared data and trainer. This will be a short run due to the small dataset and limited epochs.

In [12]:
if trainer_instance and model_for_training and train_loader:
    print("Starting T5 model fine-tuning...")
    print(f"  Max epochs: {MAX_EPOCHS}")
    print(f"  Train dataset size: {actual_train_data_size}")
    print(f"  Train batch size: {TRAIN_BATCH_SIZE}")
    if val_loader:
        print(f"  Validation dataset size: {len(val_dataset_main)}")
        print(f"  Validation batch size: {VALID_BATCH_SIZE}")
    else:
        print("  No validation dataset.")

    print(f"\nCheckpoints will be saved in: {CHECKPOINTS_DIR}")
    print(f"Logs will be saved in: {LOGS_DIR}")
    print(f"\nDevice: {EFFECTIVE_DEVICE}")
    print("This may take a few minutes depending on Colab hardware allocation...")

    start_train_time = time.time()
    try:
        trainer_instance.fit(model_for_training, train_loader, val_loader if val_loader else None)
        end_train_time = time.time()
        print(f"\n--- Training finished in {end_train_time - start_train_time:.2f} seconds ---")
        print(f"Best model checkpoint path (if val_loss monitored): {trainer_instance.checkpoint_callback.best_model_path}")
        print(f"Best model score (if val_loss monitored): {trainer_instance.checkpoint_callback.best_model_score}")
    except Exception as e:
        print(f"An error occurred during training: {e}")
        end_train_time = time.time()
        print(f"--- Training interrupted after {end_train_time - start_train_time:.2f} seconds --- ")
else:
    print("Trainer, Model, or Train Loader not available. Skipping training.")

Starting T5 model fine-tuning...
  Max epochs: 3
  Train dataset size: 44
  Train batch size: 2
  Validation dataset size: 12
  Validation batch size: 2

Checkpoints will be saved in: /content/text2program_mimic_repro/outputs/checkpoints
Logs will be saved in: /content/text2program_mimic_repro/lightning_logs

Device: cuda
This may take a few minutes depending on Colab hardware allocation...


INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name       | Type                       | Params | Mode
-----------------------------------------------------------------
0 | core_model | T5ForConditionalGeneration | 222 M  | eval
-----------------------------------------------------------------
222 M     Trainable params
0         Non-trainable params
222 M     Total params
891.614   Total estimated model params size (MB)
0         Modules in train mode
541       Modules in eval mode


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.48.0. You should pass an instance of `EncoderDecoderCache` instead, e.g. `past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`.


Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:Epoch 0, global step 22: 'val_loss' reached 0.42373 (best 0.42373), saving model to '/content/text2program_mimic_repro/outputs/checkpoints/t5-text2program-epoch=00-vlossval_loss=0.42.ckpt' as top 1


Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:Epoch 1, global step 44: 'val_loss' reached 0.04483 (best 0.04483), saving model to '/content/text2program_mimic_repro/outputs/checkpoints/t5-text2program-epoch=01-vlossval_loss=0.04.ckpt' as top 1


Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:Epoch 2, global step 66: 'val_loss' reached 0.01835 (best 0.01835), saving model to '/content/text2program_mimic_repro/outputs/checkpoints/t5-text2program-epoch=02-vlossval_loss=0.02.ckpt' as top 1
INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=3` reached.



--- Training finished in 107.10 seconds ---
Best model checkpoint path (if val_loss monitored): /content/text2program_mimic_repro/outputs/checkpoints/t5-text2program-epoch=02-vlossval_loss=0.02.ckpt
Best model score (if val_loss monitored): 0.01835285685956478


## 12. Load Best Model and Test Program Generation

Load the best checkpoint saved during training and test its ability to generate programs for a few sample questions.

In [13]:
best_checkpoint_path_found = None
loaded_model_for_inference = None
tokenizer_for_inference = text_tokenizer

if trainer_instance and hasattr(trainer_instance.checkpoint_callback, 'best_model_path') and trainer_instance.checkpoint_callback.best_model_path:
    best_checkpoint_path_found = trainer_instance.checkpoint_callback.best_model_path
    print(f"Best model checkpoint found at: {best_checkpoint_path_found}")
elif os.path.exists(CHECKPOINTS_DIR) and len(os.listdir(CHECKPOINTS_DIR)) > 0:
    # Fallback
    print(f"Trainer's best_model_path not available. Searching in {CHECKPOINTS_DIR} for any checkpoint...")
    checkpoint_files = glob.glob(str(CHECKPOINTS_DIR / "*.ckpt"))
    if checkpoint_files:
        best_checkpoint_path_found = checkpoint_files[0]
        print(f"Using fallback checkpoint: {best_checkpoint_path_found}")
    else:
        print(f"No checkpoint files found in {CHECKPOINTS_DIR}.")
else:
    print("No trainer instance or checkpoint callback information available, and no checkpoints found manually.")

if best_checkpoint_path_found and tokenizer_for_inference:
    print("\nLoading fine-tuned model from checkpoint for inference...")
    try:
        loaded_model_for_inference = T5FineTuner.load_from_checkpoint(
            best_checkpoint_path_found,
            map_location=torch.device('cpu'),
            tokenizer_instance=tokenizer_for_inference
        )
        loaded_model_for_inference.to(torch.device(EFFECTIVE_DEVICE))
        loaded_model_for_inference.eval()
        loaded_model_for_inference.ext_tokenizer = tokenizer_for_inference
        print(f"Fine-tuned model loaded successfully from {best_checkpoint_path_found} and set to evaluation mode on {EFFECTIVE_DEVICE}.")
    except Exception as e:
        print(f"Error loading model from checkpoint {best_checkpoint_path_found}: {e}")
        loaded_model_for_inference = None

    if loaded_model_for_inference:
        print("\n--- Testing Program Generation ---")
        test_questions_for_gen = [
            "How many female patients are there?", #simple q
            "What is the minimum anchor age of male patients?", # more complex
            "How many admissions were there for male patients with atrial fibrillation?" # even more complex!
        ]
        for i, q_text in enumerate(test_questions_for_gen):
            print(f"\nTest Question {i+1}: {q_text}")
            input_text_for_model = "translate English to Program: " + q_text
            input_ids_for_model = tokenizer_for_inference.encode(
                input_text_for_model,
                return_tensors='pt',
                max_length=INPUT_MAX_LEN,
                truncation=True
            ).to(torch.device(EFFECTIVE_DEVICE))

            hf_model_to_use = loaded_model_for_inference.core_model if hasattr(loaded_model_for_inference, 'core_model') else loaded_model_for_inference

            try:
                generated_ids = hf_model_to_use.generate(
                    input_ids_for_model,
                    max_length=OUTPUT_MAX_LEN + 20, # Add some additonal buffer
                    num_beams=1, # note: Paper uses beam search of 5, but 1 is faster for a quick test
                    early_stopping=True
                )
                generated_program = tokenizer_for_inference.decode(generated_ids[0], skip_special_tokens=True, clean_up_tokenization_spaces=True)
                print(f"  Generated Program: {generated_program}")
            except Exception as e:
                print(f"  Error during program generation: {e}")
else:
    print("Best checkpoint not found or tokenizer not available. Skipping program generation test.")

Best model checkpoint found at: /content/text2program_mimic_repro/outputs/checkpoints/t5-text2program-epoch=02-vlossval_loss=0.02.ckpt

Loading fine-tuned model from checkpoint for inference...
Fine-tuned model loaded successfully from /content/text2program_mimic_repro/outputs/checkpoints/t5-text2program-epoch=02-vlossval_loss=0.02.ckpt and set to evaluation mode on cuda.

--- Testing Program Generation ---

Test Question 1: How many female patients are there?


/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(


  Generated Program: count_entset(gen_entset_equal(table_name='patients', column_name='gender', value='F', entity_column='subject_id'), entity_col_name='subject_id')

Test Question 2: What is the minimum anchor age of male patients?
  Generated Program: What is the minimum anchor age of male patients?

Test Question 3: How many admissions were there for male patients with atrial fibrillation?
  Generated Program: count_entset(gen_entset_equal(table_name='patients', column_name='gender', value='M', entity_column='subject_id'), source_entity_col='subject_id')


## 13. Program Executor and Execution Accuracy (Subset)

Define a function to execute the generated program strings and compare their results against gold programs. We'll test this on a random subset of our embedded synthetic pairs for efficiency in Colab.

In [14]:
NUM_PAIRS_FOR_EXEC_ACC_SUBSET = 25 # Test on a subset for speed
y_true_labels_for_auroc = [] # To store 0 (incorrect) or 1 (correct) for AUROC
execution_accuracy_results = []

def execute_program_string_safely(program_str, question_context="N/A"):
    # Ensure db_connection and all atomic operations are in the global scope for eval
    shared_eval_globals = globals().copy()
    shared_eval_locals = {}

    try:
        # Potentially unsafe if program_str is arbitrary, but necessary for this task...
        # in a prod system, much more robust sandboxing would be required...
        result = eval(program_str, shared_eval_globals, shared_eval_locals)
        return result, True, None # result, success, error_type
    except SyntaxError as e_syn:
        # print(f"  SyntaxError executing program string: {program_str}\n    -> {e_syn}. Program string is not valid Python.")
        return None, False, "SyntaxError"
    except NameError as e_name:
        # print(f"  NameError executing program string: {program_str}\n    -> {e_name}. Likely an undefined function or variable.")
        return None, False, "NameError"
    except duckdb.BinderException as e_duck_bind:
        # print(f"  DuckDB BinderException executing program string: {program_str}\n    -> {e_duck_bind}. Problem with column/table names or types.")
        return None, False, "DuckDB.BinderException"
    except duckdb.CatalogException as e_duck_cat:
        # print(f"  DuckDB CatalogException executing program string: {program_str}\n    -> {e_duck_cat}. Usually means table/view not found.")
        return None, False, "DuckDB.CatalogException"
    except Exception as e_other:
        # print(f"  Other Exception executing program string: {program_str}\n    -> {type(e_other).__name__}: {e_other}")
        return None, False, type(e_other).__name__

def compare_execution_outcomes(gold_res, gen_res):
    # Simple comparison: exact match for numbers, or both are DataFrames and are equal (or both None)
    if isinstance(gold_res, (int, float, np.number)) and isinstance(gen_res, (int, float, np.number)):
        return gold_res == gen_res
    elif isinstance(gold_res, pd.DataFrame) and isinstance(gen_res, pd.DataFrame):
        if gold_res.empty and gen_res.empty:
             return True # Both empty is a match
        # For non-empty, sort values and reset index for robust comparison if structure is the same
        try:
            # Make sure columns are in the same order, and then sort by all columns
            gold_sorted = gold_res.sort_values(by=list(gold_res.columns)).reset_index(drop=True)
            gen_sorted = gen_res.sort_values(by=list(gen_res.columns)).reset_index(drop=True)
            return gold_sorted.equals(gen_sorted)
        except Exception:
            return False # If sorting/comparison fails, they don't match
    elif gold_res is None and gen_res is None: # Both failed to execute in a way that produced a comparable result
        return True # sidenote: This can be debated - is two failures a 'match'? For ExAcc, often yes if both are non-executable
    return False

if loaded_model_for_inference and tokenizer_for_inference and embedded_synthetic_pairs and db_connection:
    print(f"\nCalculating Execution Accuracy on a random subset of {NUM_PAIRS_FOR_EXEC_ACC_SUBSET} pairs...")

    # Ensure 'db_connection' is available in the scope for 'execute_program_string_safely' via globals()
    if 'db_connection' not in globals(): globals()['db_connection'] = db_connection

    # Select a random subset for execution accuracy calculation
    random.seed(SEED) # for reproducibility of the subset
    if len(embedded_synthetic_pairs) <= NUM_PAIRS_FOR_EXEC_ACC_SUBSET:
        subset_for_ex_acc = embedded_synthetic_pairs
    else:
        subset_for_ex_acc = random.sample(embedded_synthetic_pairs, NUM_PAIRS_FOR_EXEC_ACC_SUBSET)

    print(f"Selected {len(subset_for_ex_acc)} pairs for execution accuracy testing.")

    num_correct_executions = 0
    execution_progress_bar = tqdm(enumerate(subset_for_ex_acc), total=len(subset_for_ex_acc), desc="ExecAcc")

    for idx, pair_item in execution_progress_bar:
        question_text = pair_item['question']
        gold_program_str = pair_item['program']

        # Generate program from model
        input_text_for_model = "translate English to Program: " + question_text
        input_ids_for_model = tokenizer_for_inference.encode(input_text_for_model, return_tensors='pt', max_length=INPUT_MAX_LEN, truncation=True).to(torch.device(EFFECTIVE_DEVICE))

        hf_model_to_use_ex = loaded_model_for_inference.core_model if hasattr(loaded_model_for_inference, 'core_model') else loaded_model_for_inference
        try:
            gen_ids = hf_model_to_use_ex.generate(input_ids_for_model, max_length=OUTPUT_MAX_LEN + 20, num_beams=1, early_stopping=True)
            generated_program_str = tokenizer_for_inference.decode(gen_ids[0], skip_special_tokens=True, clean_up_tokenization_spaces=True)
        except Exception as e_gen:
            # print(f"  Error generating program for Q: {question_text_short} -> {e_gen}")
            generated_program_str = "ERROR_DURING_GENERATION"

        # Execute gold program
        gold_result, gold_success, _ = execute_program_string_safely(gold_program_str, question_text)

        # Execute generated program
        gen_result, gen_success, _ = execute_program_string_safely(generated_program_str, question_text)

        match = False
        if gold_success and gen_success:
            match = compare_execution_outcomes(gold_result, gen_result)
        elif not gold_success and not gen_success: # Both failed to execute syntactically
            match = True # Consider this a 'match' for ExAcc if both are equally broken / non-executable

        if match:
            num_correct_executions += 1
            y_true_labels_for_auroc.append(1) # Correct
        else:
            y_true_labels_for_auroc.append(0) # Incorrect

        execution_accuracy_results.append({
            'question': question_text,
            'gold_program': gold_program_str,
            'generated_program': generated_program_str,
            'gold_executed_ok': gold_success,
            'gen_executed_ok': gen_success,
            'gold_result_preview': str(gold_result)[:100] if gold_success else 'N/A',
            'gen_result_preview': str(gen_result)[:100] if gen_success else 'N/A',
            'match': match
        })
        execution_progress_bar.set_postfix({"CurrentAcc": f"{(num_correct_executions/(idx+1)):.2%}"})

    final_execution_accuracy = (num_correct_executions / len(subset_for_ex_acc)) if subset_for_ex_acc else 0
    print(f"\n--- Execution Accuracy Calculation Complete ---")
    print(f"Number of pairs tested: {len(subset_for_ex_acc)}")
    print(f"Number of correct executions (results match or both failed): {num_correct_executions}")
    print(f"Execution Accuracy (ExAcc): {final_execution_accuracy:.2%}")

    print("\nSample Execution Results (first 3):")
    for res_item in execution_accuracy_results[:3]:
        print(f"  Q: {res_item['question'][:60]}...")
        # print(f"    GoldP: {res_item['gold_program'][:60]}...")
        # print(f"    GenP:  {res_item['generated_program'][:60]}...")
        print(f"    Match: {res_item['match']}")
else:
    print("Skipping Execution Accuracy: Model, Tokenizer, Synthetic Pairs, or DB connection not available.")


Calculating Execution Accuracy on a random subset of 25 pairs...
Selected 25 pairs for execution accuracy testing.


ExecAcc:   0%|          | 0/25 [00:00<?, ?it/s]

/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:6


--- Execution Accuracy Calculation Complete ---
Number of pairs tested: 25
Number of correct executions (results match or both failed): 16
Execution Accuracy (ExAcc): 64.00%

Sample Execution Results (first 3):
  Q: How many patients had a diagnosis of 'Displaced fracture of ...
    Match: True
  Q: How many lab entries exist for Eosinophil Count (itemid 5113...
    Match: True
  Q: How many patients had a diagnosis of 'Other and unspecified ...
    Match: True


## 14. Ambiguity Detection & AUROC (Subset)

Implement ambiguity detection by generating multiple program samples for a given question, calculating the Program Inconsistency Score (PIS), and then computing the Area Under the ROC Curve (AUROC) using PIS as the prediction score and execution correctness as the true label. This also runs on a random subset.

In [15]:
NUM_SAMPLES_FOR_PIS = 5  # Number of diverse programs to generate for PIS (paper: 20)
NUM_PAIRS_FOR_AUROC_SUBSET = 25 # Must be <= NUM_PAIRS_FOR_EXEC_ACC_SUBSET if using its y_true
                                # Or recalculate y_true for this new subset
all_pis_scores_for_auroc = []
y_true_for_auroc_recalc = [] # If we recalculate for this specific subset

def generate_diverse_programs_for_auroc(question_txt, model_instance, tokenizer_instance, device_str,
                                        num_samples_to_gen=NUM_SAMPLES_FOR_PIS, top_p_val=0.9,
                                        max_len_out=OUTPUT_MAX_LEN):
    prompt = "translate English to Program: " + question_txt
    input_tok_ids = tokenizer_instance.encode(prompt, return_tensors='pt', max_length=INPUT_MAX_LEN, truncation=True).to(device_str)

    # Ensure we are using the Hugging Face model directly for generation
    hf_model = model_instance.core_model if hasattr(model_instance, 'core_model') else model_instance
    hf_model.eval() # Ensure model is in eval mode

    with torch.no_grad():
        generated_ids_list = hf_model.generate(
            input_ids=input_tok_ids,
            num_return_sequences=num_samples_to_gen,
            do_sample=True,      # Crucial for diversity
            top_p=top_p_val,     # Nucleus sampling
            top_k=0,             # Disable top_k when using top_p
            max_length=max_len_out + 20, # Buffer
            temperature=1.0      # Default temperature, can be tuned
        )
    return [tokenizer_instance.decode(g_id, skip_special_tokens=True, clean_up_tokenization_spaces=True) for g_id in generated_ids_list]

def calculate_program_inconsistency_score(program_list):
    if not program_list or len(program_list) < 2:
        return 0.0

    total_normalized_distance = 0
    num_comparisons = 0
    for i in range(len(program_list)):
        for j in range(i + 1, len(program_list)):
            prog1 = program_list[i]
            prog2 = program_list[j]
            # Levenshtein distance normalized by length of longer string
            # This gives a score between 0 (identical) and 1 (completely different)
            dist = Levenshtein.distance(prog1, prog2)
            max_len = max(len(prog1), len(prog2))
            if max_len == 0: # Both are empty strings
                normalized_dist = 0.0
            else:
                normalized_dist = dist / max_len
            total_normalized_distance += normalized_dist
            num_comparisons += 1

    return total_normalized_distance / num_comparisons if num_comparisons > 0 else 0.0

print("--- [14] Ambiguity Detection & AUROC (on a Subset) ---")
print(f"Using device: {EFFECTIVE_DEVICE}")

if loaded_model_for_inference and tokenizer_for_inference and embedded_synthetic_pairs and db_connection:
    # We need a set of questions and their ground truth correctness (y_true)
    # For consistency and speed, let's use a subset of the `execution_accuracy_results`
    # which already has questions and generated programs. We'll re-evaluate y_true for these.

    random.seed(SEED + 1) # Use a different seed or ensure subset logic is robust
    if len(embedded_synthetic_pairs) <= NUM_PAIRS_FOR_AUROC_SUBSET:
        auroc_subset_pairs = embedded_synthetic_pairs
        selected_indices_for_auroc = list(range(len(embedded_synthetic_pairs)))
    else:
        # Create a list of (original_index, pair) to track original items
        indexed_pairs = list(enumerate(embedded_synthetic_pairs))
        selected_indexed_pairs_for_auroc = random.sample(indexed_pairs, NUM_PAIRS_FOR_AUROC_SUBSET)
        auroc_subset_pairs = [item_tuple[1] for item_tuple in selected_indexed_pairs_for_auroc]
        selected_indices_for_auroc = [item_tuple[0] for item_tuple in selected_indexed_pairs_for_auroc]

    print(f"Randomly selected {len(auroc_subset_pairs)} items from {len(embedded_synthetic_pairs)} for AUROC calculation.")
    # print(f"Selected original indices for AUROC: {selected_indices_for_auroc[:5]}... (total {len(selected_indices_for_auroc)})")

    print("\nStarting PIS and y_true calculation for AUROC subset items...")
    auroc_progress_bar = tqdm(enumerate(auroc_subset_pairs), total=len(auroc_subset_pairs), desc="AUROC PIS/y_true")

    # Ensure db_connection is available globally for execute_program_string_safely
    if 'db_connection' not in globals(): globals()['db_connection'] = db_connection

    time_per_item_pis = []
    for idx_sub, current_pair_item in auroc_progress_bar:
        start_item_time = time.time()
        current_question = current_pair_item['question']
        current_gold_prog = current_pair_item['program']
        original_dataset_idx = selected_indices_for_auroc[idx_sub] # if tracking original index is needed

        # 1. Get y_true: Execute gold and a single model-generated program for correctness
        input_text_single_gen = "translate English to Program: " + current_question
        input_ids_single_gen = tokenizer_for_inference.encode(input_text_single_gen, return_tensors='pt', max_length=INPUT_MAX_LEN, truncation=True).to(torch.device(EFFECTIVE_DEVICE))

        hf_model_to_use_auroc = loaded_model_for_inference.core_model if hasattr(loaded_model_for_inference, 'core_model') else loaded_model_for_inference
        try:
            single_gen_ids = hf_model_to_use_auroc.generate(input_ids_single_gen, max_length=OUTPUT_MAX_LEN + 20, num_beams=1, early_stopping=True)
            single_generated_prog_str = tokenizer_for_inference.decode(single_gen_ids[0], skip_special_tokens=True, clean_up_tokenization_spaces=True)
        except Exception:
            single_generated_prog_str = "ERROR_DURING_SINGLE_GENERATION_FOR_YTRUE"

        gold_res_auroc, gold_ok_auroc, _ = execute_program_string_safely(current_gold_prog, current_question)
        gen_res_auroc, gen_ok_auroc, _ = execute_program_string_safely(single_generated_prog_str, current_question)

        y_true_label_current = 0
        if gold_ok_auroc and gen_ok_auroc:
            if compare_execution_outcomes(gold_res_auroc, gen_res_auroc):
                y_true_label_current = 1
        elif not gold_ok_auroc and not gen_ok_auroc:
            y_true_label_current = 1 # Both failed, considered 'correct' for this metric's purpose
        y_true_for_auroc_recalc.append(y_true_label_current)

        # 2. Calculate PIS score for the current question
        diverse_program_candidates = generate_diverse_programs_for_auroc(
            current_question, loaded_model_for_inference, tokenizer_for_inference, EFFECTIVE_DEVICE,
            num_samples_to_gen=NUM_SAMPLES_FOR_PIS, max_len_out=OUTPUT_MAX_LEN
        )
        current_pis_score = calculate_program_inconsistency_score(diverse_program_candidates)
        all_pis_scores_for_auroc.append(current_pis_score)

        item_time_taken = time.time() - start_item_time
        time_per_item_pis.append(item_time_taken)
        auroc_progress_bar.set_postfix({"PIS": f"{current_pis_score:.4f}", "y_true": y_true_label_current, "Time": f"{item_time_taken:.2f}s"})

    avg_time_pis = np.mean(time_per_item_pis) if time_per_item_pis else 0
    print(f"\nFinished PIS and y_true calculations. Average time per item: {avg_time_pis:.2f}s")

    # 3. Calculate AUROC if we have enough data
    if len(all_pis_scores_for_auroc) > 1 and len(set(y_true_for_auroc_recalc)) > 1:
        auroc_score_val = roc_auc_score(y_true_for_auroc_recalc, all_pis_scores_for_auroc)
        print(f"\nAUROC Score (using PIS): {auroc_score_val:.4f}")
        # fpr, tpr, thresholds = roc_curve(y_true_for_auroc_recalc, all_pis_scores_for_auroc)
        # TODO: Optionally plot ROC curve here using matplotlib if desired
    elif len(all_pis_scores_for_auroc) > 1 and len(set(y_true_for_auroc_recalc)) <= 1:
        print("\nAUROC calculation skipped: Only one class present in y_true values.")
        print(f"  PIS scores collected: {len(all_pis_scores_for_auroc)}")
        print(f"  y_true values collected: {y_true_for_auroc_recalc}")
    else:
        print("\nAUROC calculation skipped: Not enough data points.")
        print(f"  PIS scores collected: {len(all_pis_scores_for_auroc)}")
        print(f"  y_true values collected: {y_true_for_auroc_recalc}")
else:
    print("Skipping AUROC Calculation: Model, Tokenizer, Synthetic Pairs or DB connection not available.")

--- [14] Ambiguity Detection & AUROC (on a Subset) ---
Using device: cuda
Randomly selected 25 items from 56 for AUROC calculation.

Starting PIS and y_true calculation for AUROC subset items...


AUROC PIS/y_true:   0%|          | 0/25 [00:00<?, ?it/s]

/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:679: UserWarning: `num_beams` is set to 1. However, `early_stopping` is set to `True` -- this flag is only used in beam-based generation modes. You should set `num_beams>1` or unset `early_stopping`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/generation/configuration_utils.py:6


Finished PIS and y_true calculations. Average time per item: 3.74s

AUROC Score (using PIS): 0.2078


## 15. Clarification Prompt Example

Demonstrate how the PIS score and diverse program samples could be used to formulate a clarification prompt when the model is uncertain (high PIS).

In [16]:
print("--- Clarification Prompt Example ---")

# Pick an example question. Ideally, one that resulted in a high PIS or low confidence.
# For this demo, we'll pick one that previously showed some diversity or was problematic.
# If AUROC cell ran and `all_pis_scores_for_auroc` has items, try to pick one with high PIS.
example_question_for_clarification = "How many admissions were there for male patients with atrial fibrillation?"
use_high_pis_q = False
if 'all_pis_scores_for_auroc' in globals() and all_pis_scores_for_auroc and 'auroc_subset_pairs' in globals():
    if len(all_pis_scores_for_auroc) == len(auroc_subset_pairs):
        # Try to find a question with high PIS from the AUROC subset
        pis_q_pairs = sorted(zip(all_pis_scores_for_auroc, auroc_subset_pairs), key=lambda x: x[0], reverse=True)
        if pis_q_pairs and pis_q_pairs[0][0] > 0.1: # If highest PIS is somewhat significant
            example_question_for_clarification = pis_q_pairs[0][1]['question']
            use_high_pis_q = True
            print(f"Using a question with high PIS ({pis_q_pairs[0][0]:.4f}) for clarification example.")

print(f"Example Question: \"{example_question_for_clarification}\"")

if loaded_model_for_inference and tokenizer_for_inference:
    clarify_diverse_programs = generate_diverse_programs_for_auroc(
        example_question_for_clarification,
        loaded_model_for_inference,
        tokenizer_for_inference,
        EFFECTIVE_DEVICE,
        num_samples_to_gen=NUM_SAMPLES_FOR_PIS, # Use same number of samples as for PIS calc
        max_len_out=OUTPUT_MAX_LEN
    )
    print(f"\nGenerated {len(clarify_diverse_programs)} diverse program samples:")
    unique_clarify_programs = sorted(list(set(clarify_diverse_programs)))
    for i_cp, cp_text in enumerate(unique_clarify_programs[:5]): # Show up to 5 unique ones
        print(f"  Sample {i_cp+1}: {cp_text}")

    clarify_pis_score = calculate_program_inconsistency_score(clarify_diverse_programs)
    print(f"\nProgram Inconsistency Score (PIS) for these {len(clarify_diverse_programs)} samples: {clarify_pis_score:.4f}")
    print(f"Number of unique programs generated: {len(unique_clarify_programs)}")

    # Simulate Clarification Prompt
    print("\n--- Simulated Clarification Prompt ---")
    if clarify_pis_score > 0.1 and len(unique_clarify_programs) > 1: # Threshold for when to ask
        print("I found a few ways to interpret your question. Could you clarify?")
        print("For example, I see these possible interpretations based on the programs I could generate:")
        # Show a couple of the most different interpretations (hard to quantify easily without semantic understanding)
        # For now, just show the first 2-3 unique ones if they exist.
        for i_prompt, prog_text in enumerate(unique_clarify_programs[:min(3, len(unique_clarify_programs))]):
            # We'd need a way to summarize the program's intent or show the SQL if it's a SQL-based program
            # For now, just indicate it leads to a certain program string
            print(f"   Interpretation {i_prompt+1}: Perform operations leading to '{prog_text[:80]}...'?")
    elif len(unique_clarify_programs) == 1:
        print(f"Based on your question, I would execute: {unique_clarify_programs[0]}")
    else: # Low PIS or no unique programs
        # Could still show the top program if confident
        if unique_clarify_programs:
            print(f"Based on your question, I'm fairly confident the program should be: {unique_clarify_programs[0]}")
        else:
            print("I'm having trouble generating a confident program for your question.")
else:
    print("Skipping Clarification Prompt Example: Model or Tokenizer not available.")

--- Clarification Prompt Example ---
Using a question with high PIS (0.6921) for clarification example.
Example Question: "What is the maximum anchor age of any patient?"

Generated 5 diverse program samples:
  Sample 1: Gibt das maximale Ankeralter eines Patienten fest?
  Sample 2: What is the maximum anchor age of any patient?
  Sample 3: nderung des Antennalts eines Patienten?
  Sample 4: welches ankeralter maximal ein Patient erreicht?

Program Inconsistency Score (PIS) for these 5 samples: 0.6209
Number of unique programs generated: 4

--- Simulated Clarification Prompt ---
I found a few ways to interpret your question. Could you clarify?
For example, I see these possible interpretations based on the programs I could generate:
   Interpretation 1: Perform operations leading to 'Gibt das maximale Ankeralter eines Patienten fest?...'?
   Interpretation 2: Perform operations leading to 'What is the maximum anchor age of any patient?...'?
   Interpretation 3: Perform operations leadin

## 16. Conclusion and Next Steps

This notebook successfully demonstrated a simplified, Colab-friendly pipeline for reproducing key aspects of the "Uncertainty-Aware Text-to-Program" paper. It utilized the MIMIC-IV demo dataset (v2.2) downloaded directly in the environment and a pre-defined set of 56 embedded synthetic Question-Program pairs for model training and evaluation. The process was executed end-to-end, including data setup, model fine-tuning on a GPU, and evaluation of both program generation accuracy and uncertainty metrics.

**Key Achievements:**

*   Successfully set up an in-memory DuckDB instance with MIMIC-IV demo CSVs, enabling SQL-like queries via Python.
*   Implemented and tested a suite of 14 atomic database operations as Python functions.
*   Fine-tuned a T5-base model for 3 epochs on the 56 embedded synthetic Question-Program pairs using PyTorch Lightning on a Colab GPU (T4).
*   Evaluated the fine-tuned model's program generation capabilities, achieving an Execution Accuracy (ExAcc) of **64.00%** on a random subset of 25 test pairs. This score, while modest, indicates that the model learned basic translation patterns even from the very small dataset.
*   Demonstrated ambiguity detection by calculating the Program Inconsistency Score (PIS) from 5 diverse samples per question.
*   Calculated an Area Under the ROC Curve (AUROC) of **0.2078** for PIS as a predictor of program correctness on a 25-item subset. This low score suggests that PIS, in this specific configuration (small N_samples, severely under-trained model), was not a reliable positive indicator of correctness and, in fact, showed a negative correlation.
*   Showcased a clarification prompt mechanism that successfully utilized a high PIS score to flag a question where the model was highly uncertain. This uncertainty manifested as the model generating outputs in other languages (e.g., German in this case) or merely repeating the input, underscoring the PIS's utility in identifying such confused states.

**Limitations & Future Work:**

*   **Primary Limitation - Dataset Size:** The model was trained on only 56 synthetic pairs. This is significantly smaller than the ~10k pairs used in the original paper. Performance in terms of generalization, handling complex queries, and the reliability of uncertainty metrics (like PIS/AUROC) is heavily constrained by this.
*   **Demo Data Constraints:** While the MIMIC-IV demo data allowed for end-to-end execution, its small scale and limited complexity might not fully reflect the challenges or opportunities present in the full dataset.
*   **Simplified Uncertainty & Evaluation:**
    *   The AUROC for PIS was low, indicating PIS was not effective for this under-trained model. More training data, a larger number of samples for PIS calculation (e.g., N=20 as in the paper), or more advanced uncertainty techniques (like model ensembling) would likely be needed to achieve a more predictive uncertainty score.
    *   Execution accuracy was based on a 25-item subset; broader evaluation would be beneficial.
*   **Generation Quality & Error Analysis:**
    *   The model exhibited common failure modes of under-trained text-to-SQL/program models, such as repeating the input or generating incomplete/syntactically incorrect programs for more complex queries.
    *   The observed generation of non-English text (German) for an English input, particularly when PIS was high, is an interesting artifact of using a multilingual base model (T5) with insufficient task-specific fine-tuning. This warrants further investigation into controlling output language and managing model confusion.
*   **Atomic Operations and Program Space:** The set of atomic operations, while functional, could be expanded to cover more complex SQL-like capabilities, which in turn would require more sophisticated synthetic data generation.

This reproduction provides a practical, foundational pipeline and valuable insights into the challenges and behaviors of training text-to-program models with limited data. For a more faithful and robust reproduction that achieves higher performance and more reliable uncertainty estimation, significantly more diverse training data (ideally ~10k+ pairs generated from the full MIMIC-IV dataset) and potentially more extensive hyperparameter tuning and training epochs would be essential.